File: eclipse-initial/tests/native/org.eclipse.ice.reactor.native.test/FeatureSetTester.cpp
/*******************************************************************************
 * Copyright (c) 2012, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE FeatureSetTester_testSuite
#include <boost/test/included/unit_test.hpp>
#include <FeatureSet.h>
#include <vector>
#include <memory>
#include <string>
#include <IData.h>

using namespace ICE_Reactor;

BOOST_AUTO_TEST_SUITE(FeatureSetTester_testSuite)



BOOST_AUTO_TEST_CASE(checkConstruction) {

    // begin-user-code

    //Local Declarations
    std::string feature = "pin power";
    std::string nullFeature;
    std::string spaceFeature = " ";
    std::string emptyFeature;

    //Test normal construction
    std::unique_ptr<FeatureSet> featureSet1 (new FeatureSet(feature));
    BOOST_REQUIRE_EQUAL(feature, featureSet1.get()->getName());

    //Invalid construction - null
    std::unique_ptr<FeatureSet> featureSet2 (new FeatureSet(nullFeature));
    BOOST_REQUIRE_EQUAL(emptyFeature, featureSet2.get()->getName()); //Defaults to empty

    //Invalid construction - space
    std::unique_ptr<FeatureSet> featureSet3 (new FeatureSet(spaceFeature));
    BOOST_REQUIRE_EQUAL(emptyFeature, featureSet3.get()->getName()); //Set as empty!

    return;

    // end-user-code

}
BOOST_AUTO_TEST_CASE(checkData) {

    // begin-user-code

    //Local Declarations
    std::string feature1 = "Feature 1";
    std::string feature2 = "Feature 2";
    std::string nullFeature;
    std::string emptyFeature = " ";

    //Setup Features
    std::shared_ptr<FeatureSet> set1 (new FeatureSet(feature1));
    std::shared_ptr<FeatureSet> set2 (new FeatureSet(nullFeature));
    std::shared_ptr<FeatureSet> set3 (new FeatureSet(emptyFeature));

    //Setup data
    std::shared_ptr<LWRData> data1 (new LWRData(feature1));
    std::shared_ptr<LWRData> data2 (new LWRData(feature1));
    std::shared_ptr<LWRData> data3 (new LWRData(feature2));

    //Try to add data to the feature set
    BOOST_REQUIRE_EQUAL(set1.get()->addIData(data1), true);
    //Check contents
    BOOST_REQUIRE_EQUAL(1, set1.get()->getIData().size());
    BOOST_REQUIRE_EQUAL(data1.get()->operator==(*(set1->getIData().at(0).get())), true);

    //Try to add data to the feature set
    BOOST_REQUIRE_EQUAL(set1.get()->addIData(data2), true);
    //Check contents
    BOOST_REQUIRE_EQUAL(2, set1.get()->getIData().size());
    BOOST_REQUIRE_EQUAL(data1.get()->operator==(*(set1.get()->getIData().at(0).get())), true);
    BOOST_REQUIRE_EQUAL(data2.get()->operator==(*(set1.get()->getIData().at(1).get())), true);

    //Try to add data to the feature set - invalid data
    BOOST_REQUIRE_EQUAL(set1.get()->addIData(data3), false);
    //Check contents - show that only two pieces remain
    BOOST_REQUIRE_EQUAL(2, set1.get()->getIData().size());
    BOOST_REQUIRE_EQUAL(data1.get()->operator==(*(set1.get()->getIData().at(0).get())), true);
    BOOST_REQUIRE_EQUAL(data2.get()->operator==(*(set1.get()->getIData().at(0).get())), true);


    //Check removal
    BOOST_REQUIRE_EQUAL(set1.get()->removeIDataAtIndex(5), false);
    BOOST_REQUIRE_EQUAL(set1.get()->removeIDataAtIndex(-1), false);
    BOOST_REQUIRE_EQUAL(set1.get()->removeIDataAtIndex(1), true);

    //Check size
    BOOST_REQUIRE_EQUAL(1, set1.get()->getIData().size());

    //Add it back
    BOOST_REQUIRE_EQUAL(true, set1.get()->addIData(data2));


    //Show that data can only be removed from the feature set by manipulating the Feature calls directly
    std::vector< std::shared_ptr<IData> > dataList = set1.get()->getIData();
    dataList.erase(dataList.begin());
    //Check contents - show that the item was not removed
    BOOST_REQUIRE_EQUAL(2, set1.get()->getIData().size());
    BOOST_REQUIRE_EQUAL(data1.get()->operator==(*(set1.get()->getIData().at(0).get())), true);

    //Try to setup bad FeatureSets and show that data can not be added

    //Show the behavior of a null feature set
    BOOST_REQUIRE_EQUAL(set2.get()->addIData(data1), false);
    BOOST_REQUIRE_EQUAL(0, set2.get()->getIData().size());
    dataList = set2.get()->getIData();
    dataList.push_back(data1);

    //Show the list can not be manipulated
    BOOST_REQUIRE_EQUAL(0, set2.get()->getIData().size());

    //Show the behavior of a empty string feature set
    BOOST_REQUIRE_EQUAL(set3.get()->addIData(data1), false);
    BOOST_REQUIRE_EQUAL(0, set3.get()->getIData().size());
    dataList = set3.get()->getIData();
    dataList.push_back(data1);

    //Show the list can not be manipulated
    BOOST_REQUIRE_EQUAL(0, set3.get()->getIData().size());

    return;

    // end-user-code

}
BOOST_AUTO_TEST_CASE(checkEquality) {

    // begin-user-code

    // Create an Object
    FeatureSet object("Feature 1");

    // Set its data

    std::shared_ptr<LWRData> data1(new LWRData("Feature 1"));
    std::shared_ptr<LWRData> data2(new LWRData("Feature 1"));

    object.addIData(data1);
    object.addIData(data2);

    // Create another FeatureSet to assert Equality with the last
    FeatureSet equalObject("Feature 1");

    // Set its data, equal to object
    for(int i = 0; i < object.getIData().size(); i++) {
        equalObject.addIData(object.getIData().at(i));
    }

    // Create a unEqualObject that is not equal to Object
    FeatureSet unEqualObject("Feature 1");

    // Set its data, not equal to testLWRComponent
    unEqualObject.addIData(equalObject.getIData().at(0));

    // Create a third object to test Transitivity
    FeatureSet transitiveObject("Feature 1");

    //Set it equal
    // Set its data, equal to object
    for(int i = 0; i < object.getIData().size(); i++) {
        transitiveObject.addIData(object.getIData().at(i));
    }

    // Assert that these two LWRComponents are equal
    BOOST_REQUIRE_EQUAL(object==(equalObject), true);

    // Assert that two unequal objects returns false
    BOOST_REQUIRE_EQUAL(object==(unEqualObject), false);

    // Check that equals() is Reflexive
    // x==(x) = true
    BOOST_REQUIRE_EQUAL(object==(object), true);

    // Check that equals() is Symmetric
    // x==(y) = true iff y==(x) = true
    BOOST_REQUIRE_EQUAL(object==(equalObject) && equalObject==(object), true);

    // Check that equals() is Transitive
    // x==(y) = true, y==(z) = true => x==(z) = true
    if (object==(equalObject) && equalObject==(transitiveObject)) {
        BOOST_REQUIRE_EQUAL(object==(transitiveObject), true);
    } else {
        BOOST_FAIL("Equality check failed.  Exiting.");
    }

    // Check the Consistent nature of equals()
    BOOST_REQUIRE_EQUAL(object==(equalObject)
                        && object==(equalObject)
                        && object==(equalObject), true);
    BOOST_REQUIRE_EQUAL(!(object==(unEqualObject))
                        && !(object==(unEqualObject))
                        && !(object==(unEqualObject)), true);

    return;

    // end-user-code

}
BOOST_AUTO_TEST_CASE(checkCopying) {

    // begin-user-code

    // Create an Object
    FeatureSet object("Feature 2");
    std::shared_ptr<FeatureSet> clonedObject;

    // Set its data
    std::shared_ptr<LWRData> data1(new LWRData("Feature 2"));
    std::shared_ptr<LWRData> data2(new LWRData("Feature 2"));

    //Add data
    object.addIData(data1);
    object.addIData(data2);

    //Run the copy routine
    FeatureSet copyObject(object);

    //Check contents
    BOOST_REQUIRE_EQUAL(object==(copyObject), true);

    //Run the clone routine
    clonedObject = object.clone();

    //Remove reference
    FeatureSet clonedObjectRef = *clonedObject.get();

    //Check contents
    BOOST_REQUIRE_EQUAL(object==clonedObjectRef, true);

    return;

    // end-user-code

}
BOOST_AUTO_TEST_SUITE_END()

File: tests/org.eclipse.ice.reactor.test/src/org/eclipse/ice/reactor/test/FeatureSetTester.java
/*******************************************************************************
 * Copyright (c) 2013, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
package org.eclipse.ice.reactor.test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.util.ArrayList;

import org.eclipse.ice.analysistool.IData;
import org.eclipse.ice.reactor.FeatureSet;
import org.eclipse.ice.reactor.LWRData;
import org.junit.Test;

/**
 * <p>
 * A class that tests the operations on FeatureSet.
 * </p>
 * 
 * @author Scott Forest Hull II
 */
public class FeatureSetTester {
	/**
	 * <p>
	 * An operation that checks the constructor and getName operation.
	 * </p>
	 * 
	 */
	@Test
	public void checkConstruction() {

		// Local Declarations
		FeatureSet featureSet;
		String feature = "pin power";
		String nullFeature = null;
		String emptyFeature = " ";

		// Test normal construction
		featureSet = new FeatureSet(feature);
		assertEquals(feature, featureSet.getName());

		// Invalid construction - null
		featureSet = new FeatureSet(nullFeature);
		assertEquals(nullFeature, featureSet.getName());

		// Invalid construction - empty
		featureSet = new FeatureSet(emptyFeature);
		assertEquals(nullFeature, featureSet.getName()); // Set as null!

	}

	/**
	 * <p>
	 * An operation that checks the getIData and addIData operation.
	 * </p>
	 * 
	 */

	@Test
	public void checkData() {

		// Local Declarations
		String feature1 = "Feature 1";
		String feature2 = "Feature 2";
		String nullFeature = null;
		String emptyFeature = " ";
		ArrayList<IData> dataList;

		// Setup Features
		FeatureSet set1 = new FeatureSet(feature1);
		FeatureSet set2 = new FeatureSet(nullFeature);
		FeatureSet set3 = new FeatureSet(emptyFeature);

		// Setup data
		LWRData data1 = new LWRData(feature1);
		LWRData data2 = new LWRData(feature1);
		LWRData data3 = new LWRData(feature2);

		// Try to add data to the feature set
		assertTrue(set1.addIData(data1));
		// Check contents
		assertEquals(1, set1.getIData().size());
		assertTrue(data1.equals(set1.getIData().get(0)));

		// Try to add data to the feature set
		assertTrue(set1.addIData(data2));
		// Check contents
		assertEquals(2, set1.getIData().size());
		assertTrue(data1.equals(set1.getIData().get(0)));
		assertTrue(data2.equals(set1.getIData().get(0)));

		// Try to add data to the feature set - invalid data
		assertFalse(set1.addIData(data3));
		// Check contents - show that only two pieces remain
		assertEquals(2, set1.getIData().size());
		assertTrue(data1.equals(set1.getIData().get(0)));
		assertTrue(data2.equals(set1.getIData().get(0)));

		// Show that data can be removed from the feature set by manipulating
		// the arraylist directly
		dataList = set1.getIData();
		dataList.remove(1);
		// Check contents - show that the item was removed
		assertEquals(1, set1.getIData().size());
		assertTrue(data1.equals(set1.getIData().get(0)));

		// Try to setup bad FeatureSets and show that data can not be added

		// Show the behavior of a null feature set
		assertFalse(set2.addIData(data1));
		assertEquals(0, set2.getIData().size());
		dataList = set2.getIData();
		dataList.add(data1);

		// Show the list can not be manipulated
		assertEquals(0, set2.getIData().size());

		// Show the behavior of a empty string feature set
		assertFalse(set3.addIData(data1));
		assertEquals(0, set3.getIData().size());
		dataList = set3.getIData();
		dataList.add(data1);

		// Show the list can not be manipulated
		assertEquals(0, set3.getIData().size());

	}

	/**
	 * <p>
	 * An operation that checks the equality operations.
	 * </p>
	 * 
	 */
	@Test
	public void checkEquality() {

		// Create an Object
		FeatureSet object = new FeatureSet("Feature 1");

		// Set its data
		object.addIData(new LWRData("Feature 1"));
		object.addIData(new LWRData("Feature 1"));

		// Create another FeatureSet to assert Equality with the last
		FeatureSet equalObject = new FeatureSet("Feature 1");

		// Set its data, equal to object
		for (int i = 0; i < object.getIData().size(); i++) {
			equalObject.addIData(object.getIData().get(i));
		}

		// Create a unEqualObject that is not equal to Object
		FeatureSet unEqualObject = new FeatureSet("Feature 1");

		// Set its data, not equal to testLWRComponent
		unEqualObject.addIData(equalObject.getIData().get(0));

		// Create a third object to test Transitivity
		FeatureSet transitiveObject = new FeatureSet("Feature 1");

		// Set it equal
		// Set its data, equal to object
		for (int i = 0; i < object.getIData().size(); i++) {
			transitiveObject.addIData(object.getIData().get(i));
		}

		// Assert that these two LWRComponents are equal
		assertTrue(object.equals(equalObject));

		// Assert that two unequal objects returns false
		assertFalse(object.equals(unEqualObject));

		// Check that equals() is Reflexive
		// x.equals(x) = true
		assertTrue(object.equals(object));

		// Check that equals() is Symmetric
		// x.equals(y) = true iff y.equals(x) = true
		assertTrue(object.equals(equalObject) && equalObject.equals(object));

		// Check that equals() is Transitive
		// x.equals(y) = true, y.equals(z) = true => x.equals(z) = true
		if (object.equals(equalObject) && equalObject.equals(transitiveObject)) {
			assertTrue(object.equals(transitiveObject));
		} else {
			fail();
		}

		// Check the Consistent nature of equals()
		assertTrue(object.equals(equalObject) && object.equals(equalObject)
				&& object.equals(equalObject));
		assertTrue(!object.equals(unEqualObject)
				&& !object.equals(unEqualObject)
				&& !object.equals(unEqualObject));

		// Assert checking equality with null value returns false
		assertFalse(object==null);

		// Assert that two equal objects have the same hashcode
		assertTrue(object.equals(equalObject)
				&& object.hashCode() == equalObject.hashCode());

		// Assert that hashcode is consistent
		assertTrue(object.hashCode() == object.hashCode());

		// Assert that hashcodes are different for unequal objects
		assertFalse(object.hashCode() == unEqualObject.hashCode());

	}

	/**
	 * <p>
	 * An operation that checks the operations on copying routines.
	 * </p>
	 * 
	 */

	@Test
	public void checkCopying() {

		// Create an Object
		FeatureSet object = new FeatureSet("Feature 1");
		FeatureSet copyObject, clonedObject;

		// Set its data
		object.addIData(new LWRData("Feature 1"));
		object.addIData(new LWRData("Feature 1"));

		// Run the copy routine
		copyObject = new FeatureSet(null);
		copyObject.copy(object);

		// Check contents
		assertTrue(object.equals(copyObject));

		// Run the clone routine
		clonedObject = (FeatureSet) object.clone();

		// Check contents
		assertTrue(object.equals(clonedObject));

		// Pass null for the copy routine
		copyObject.copy(null);

		// Show that nothing as changed
		assertTrue(object.equals(copyObject));

	}
}